### ISSUE TITLE
Add methods for getting the current time in a specific time zone

### ISSUE THREAD

**Issue by mj1856** [dotnet/corefx#25089]
We presently have `DateTime.Now` and `DateTimeOffset.Now` for getting the current time in the computer's local time zone.  We also have `DateTime.UtcNow` and `DateTimeOffset.UtcNow` for getting the current time in UTC.  However, we don't have a property or method to get the time in a *specific* time zone.  To achieve that, one currently has to first get the UTC time, then get a specific time zone (if one doesn't have it already), and then convert the time to that time zone.  This proposal is to add an API to make that one step instead of two or three.

## Proposed API
```diff
namespace System {
     public class TimeZoneInfo {
+        public DateTimeOffset Now { get; }
     }
 }
```

## Example Usage
```csharp
DateTimeOffset now = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time").Now;
```

Note that the above has been edited based on the conversation in this thread.  The original proposal was to add `NowInTimeZone` methods for `DateTime` and `DateTimeOffset`.

Other alternatives to consider would be to have `.GetCurrentTime()` as a method instead of the `.Now` property, or to introduce a `ZonedDateTime` type.

**Comment by Clockwork-Muse** [comment]
Not against this, but... possibility this strengthens the impression `DateTime`/`Offset` will respect the zone?

**Comment by mj1856** [comment]
It wouldn't change anything about the data types themselves, so I don't think so.  Also, I think if the Intellisense that pops up when one starts typing `DateTime.Now` offers `NowInTimeZone` should at least get that developer thinking about the time zone.

For example, in ASP.Net applications, all too often I find folks seem to think that `DateTime.Now` should return their user's local time, in the same way that the `CurrentCulture` can represent their user's culture.  This cannot happen because there's no time zone information passed over HTTP headers, so `DateTime.Now` represents the server's time zone instead.

And yes, I'm well aware that `DateTime.UtcNow` or using `DateTimeOffset` is usually a better choice for web apps, but there are still indeed many use cases for local time in the user's time zone.

Ultimately, this would remove the need for boiler-plate code like the following:
```csharp
DateTimeOffset utc = DateTimeOffset.UtcNow;
TimeZoneInfo tz = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time");
DateTimeOffset eastern = TimeZoneInfo.ConvertTime(utc, tz);
```
or the `DateTime` version:
```csharp
DateTime utc = DateTime.UtcNow;
TimeZoneInfo tz = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time");
DateTime eastern = TimeZoneInfo.ConvertTimeFromUtc(utc, tz);
```

Additionally, I think it would prevent common errors.  For example, I sometimes see code like this:
```csharp
DateTime now = DateTime.Now;
TimeZoneInfo tz = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time");
TimeSpan offset = tz.BaseUtcOffset;
DateTime eastern = now.ToUniversalTime() + offset;
if (now.IsDaylightSavingTime())
    eastern = eastern.AddHours(1);
```
... which has lots of subtle mistakes that could return correct results in some time zones and incorrect results in others, or return correct results one day, and incorrect results the next (due to DST dates/times being different around the world).

**Comment by joperezr** [comment]
cc: @tarekgh

**Comment by joperezr** [comment]
Personally, I don't think this Api will give that much value given that the workaround is writing three lines as opposed to 1, but I would like to see what @tarekgh thinks.

**Comment by tarekgh** [comment]
@mj1856 I have suggested in the issue https://github.com/dotnet/corefx/issues/25110 to have a new type ZonedDateTime . What you think instead of proposing all these APIs to just include them in ZonedDateTime? I mean we'll have ZonedDateTime.Now?

**Comment by tarekgh** [comment]
@mj1856 if you want to discuss all proposals offline, that would be more effective. we can set up some time with @joperezr and go through these.

**Comment by mj1856** [comment]
No, I do not think we should have a `ZonedDateTime` type, for reasons previously discussed.

In general, I've been trying to follow precedent and established guidelines for API proposals.  Others, such as #24449 have followed this form.  I'm not sure why this is different.  Clearly, setting up the corefxlab project for `System.Time` was not effective, as it has sat idle with low visibility for over two years.  My intent now is to decompose that work into individual API proposals, as these seem to actually get traction.

I'll be happy to set up some time to discuss with you in more detail if you think that would be useful.

**Comment by mj1856** [comment]
@joperezr - Those three lines are often written incorrectly.  See my example above.

The larger problem is that many developers do not think about the concern of time zones up front.  The naming of these APIs are significant such that they appear nearby alphabetically in Intellisense when one reaches for `DateTime.Now`.

**Comment by tarekgh** [comment]
@mj1856 I am not proposing to use corefxlab. in general I don't like bloating DateTime/Offset with the TZI functionality. you may think to add the functionality in TZI instead if you like.

>  I do not think we should have a ZonedDateTime type, for reasons previously discussed.

I don't recall the details. could you summarize why you don't like teh idea of ZonedDateTime?

**Comment by mj1856** [comment]
Sure. To reiterate the reasons previously discussed for not making a `System.ZonedDateTime` type:

- Like `DateTimeOffset`, a `ZonedDateTime` has two aspects - the instant along the universal timeline, and the local time representation at that instant.  However, it also has a reference to a time zone, which makes it difficult to implement `IComparable` for reliable sorting.

  For example, how does the value `2017-11-28T00:00:00-08:00[America/Los_Angeles]` compare to the value `2017-11-28T00:00:00-08:00[America/Vancouver]`?  They aren't equal because they have different time zone ids, but one isn't greater or less than each other with regard to either the local time or the corresponding universal time.

- There is no standardized single-string representation (ISO or otherwise) of such a type.  The bracket notation I just used is used in Java, but Noda Time uses a different notation.  There is a discussion about this [on StackOverflow here](https://stackoverflow.com/a/42196445/634824).

- Time zone IDs in .NET are platform specific (IANA vs Windows).  [My TimeZoneConverter project](https://github.com/mj1856/TimeZoneConverter) can be used to work around that, but the serialization issue would basically make using it a hard requirement for interoperability.  If not, then you'd start seeing public APIs giving results like `2018-06-01T00:00:00+02:00[Romance Standard Time]`, which would push the issue of time zone conversion out to many other languages.  I'd like to keep it contained within the .Net ecosystem as much as possible.

- When using `ZonedDateTime` from Noda Time (and in Java), many application developers find that `ZonedDateTime` is most useful as an intermediate layer for time zone conversions between `Instant` and `LocalDateTime`, and less useful as a "primary" type.

  Consider that with both timestamping (recording the "now" that something is happening), and evaluating past timestamps, only a single offset applies, so one has everything they need with `DateTimeOffset`.  Scenarios that require *future* information, such as scheduling, do require time zone information. However, a recurrence pattern typically owns the time zone separate from any given timestamp.

  Really the only direct action that one might take on a `ZonedDateTime` would be to convert to another time zone (which we already have with `TimeZoneInfo.ConverTime`, or to add/subtract units of time in that particular time zone (which can hang off existing objects, as I've proposed in #25110).  IMHO, it doesn't make sense to build another type just to be a container for these methods.

- There aren't a lot of integration points for a `ZonedDateTime` type within other parts of the ecosystem.  Contrast that with `Date`, and `Time` types that I've been working on in corefxlab, which already have corresponding types in SQL, XSD, OData, and others.

- If one really feels like they *want* a `ZonedDateTime` type to model their data, they can always create their own, or pull in Noda Time.

- We had previously discussed limiting the number of new types introduced, such that we address the primary pain points and make them feel natural within our current model.  The two (`Date` and `Time`) types make a lot of sense.  This would be a third, and IMHO would raise more questions than it answered.

**Comment by tarekgh** [comment]
@mj1856 that is enough concerns :-) 

how do you feel to have the requested functionality to be on TZI instead of DateTime/Offset? As I mentioned, I am trying to avoid bloating the DateTime with the TZI functionality while DateTime calculations itself doesn't really care about the TZ. in other words, I want to have the guidelines as if you need to do anything that is TZ related, do it in TZI.

**Comment by mj1856** [comment]
Sure, they could land there, but I'm not sure that they would be as discoverable.  There's a real benefit to the Intellisense factor.

Also, we already *do* time zone related things, just only for the local time zone and UTC. ;)

Would you feel better about them landing as extension methods?  If so, what class would be appropriate to place them in?

**Comment by tarekgh** [comment]
> but I'm not sure that they would be as discoverable

Why do you think it will be discoverable in DateTime and not in TZI? If I am thinking to get the time now in any time zone, I would look at the TZI (as logical thinking).

> Also, we already do time zone related things, just only for the local time zone and UTC

Right but this is already a problem and we don't want to make it bigger. When people see TZ related functionality they feel DateTime/Offset is handling the TZ calculation which is not true. 

> Would you feel better about them landing as extension methods? If so, what class would be appropriate to place them in?

I don't think extension methods to DateTime/Offset will make any difference than putting it inside the actual structs. Also, I don't think you can write these static methods as extension methods.

**Comment by mj1856** [comment]
By discoverable, I mean that there is a very strong habit for one to type `DateTime.Now` and never think about time zones.  Moreso, there's a tendency for one to think that time zone is somehow handled automagically in the way that culture is - especially in web applications.  See [here](https://stackoverflow.com/q/33839556/634824), [here](https://stackoverflow.com/q/19241815/634824) and [here](https://stackoverflow.com/q/8589014/634824) for a few examples.

I was hoping that by making the method called `DateTime.NowInTimeZone`, that when a novice developer started typing .-N-o-w, they would then see an alternative and get them thinking in the right direction.

(Good point about them being static.  Momentary lapse... )

**Comment by tarekgh** [comment]
> By discoverable, I mean that there is a very strong habit for one to type DateTime.Now and never think about time zones. Moreso, there's a tendency for one to think that time zone is somehow handled automagically in the way that culture is - especially in web applications. See here, here and here for a few examples.

That is exactly my point. I want to get people not to think DateTime is handling TZ. Adding the suggested APIs to DateTime will make people think more this is the case.

Although I agree with you on the discoverability now,  I am seeing TimeZoneInfo.Now make a lot of sense. If someone asks how I can get the time Now on any time zone, the answer would be, call Now property on the TZ object. This is better than answering it as call the DateTime API and pass the TZ object/ID to get it.

**Comment by Clockwork-Muse** [comment]
>When using `ZonedDateTime` from Noda Time (and in Java), many application developers find that `ZonedDateTime` is most useful as an intermediate layer for time zone conversions between Instant and `LocalDateTime`, and less useful as a "primary" type.

In Java you can go from `Instant` to `LocalDateTime` (and back) directly, assuming you have a zone.  Don't know if that means more people are skipping it.  Personally, I'd probably try to get a `Zoned` as soon as possible, and hold onto it as long as I can, instead of a `Local` (and having to pass around an associated zone), and just ignoring the zone for most formatting situations.

>There aren't a lot of integration points for a `ZonedDateTime` type within other parts of the ecosystem. Contrast that with Date, and Time types that I've been working on in corefxlab, which already have corresponding types in HTML5, SQL, XSD, OData, and others.

Contrast that with an `Instant`, which is semantically what most of the system should be using (in place of `DateTime` or `DateTimeOffset`).  But it's probably too late to get that added now (given API breakage).

>If one really feels like they want a `ZonedDateTime` type to model their data, they can always create their own, or pull in Noda Time.

NodaTime only lets you add time-based values, though, which irks me (I get the reasoning behind the decision, I'm not sure I would've made the same one).  Which is possibly also why people may be skipping over it.  Creating your own type is fraught with hidden perils, if people are even aware there's a difference between `Zoned` and `Offset` (which, granted, likely wouldn't help them choose the proper type if we had them, :sigh:).

**Comment by mj1856** [comment]
@tarekgh - so, if I'm understanding correctly, you'd prefer:

## Proposed API
```diff
namespace System {
     public class TimeZoneInfo {
+        public static DateTimeOffset Now { get; }
     }
 }
```

## Example Usage
```csharp
DateTimeOffset now = TimeZoneInfo.FindSystemTimeZoneById("Eastern Standard Time").Now;
```

I suppose I'm ok with this.  It's simpler, for sure.  It may not help with the Intellisense angle, but at least it's easy to talk about.

Any thoughts about making it a method like `.GetCurrentTime()`?  I have heard feedback about how `.Now`/`.UtcNow` should have been methods since their value changes with each read.  Or are we just going for consistency with the existing properties?

**Comment by tarekgh** [comment]
> Any thoughts about making it a method like .GetCurrentTime()? 

I am fine either way, even I prefer the property. you may list this as an alternative option in the proposal and let the design committee advise about that.

**Comment by mj1856** [comment]
Updated the issue summary at the top with the changes.  Thanks.

**Comment by tarekgh** [comment]
Thanks @mj1856

@joperezr please have a look and let's know if you have any feedback. otherwise, you may mark this as ready for design review.

**Comment by mj1856** [comment]
After thinking this through some more, I'm not *completely* opposed to a `ZonedDateTime` type.  @tarekgh, @joperezr  I'll discuss with you both offline.

**Comment by mj1856** [comment]
After meeting with @tarekgh, we've decided to close this for now.  We will discuss this scenario as part of a larger effort, which may or may not involve a ZonedDateTime type.
